checkers/src/Pawn.cpp

```diff
@@ -1,96 +1,81 @@
+                    if (current.isInitial()) current = tmp;
+                    else current = current.merge(tmp);
+                    board.makeMove(tmp);
+                    if (newP.isLastRow(isUser())) {
+                        King k(newP, isWhite(), isUser(), board);
+                        auto kingMoves = k.getValidMoves(board, true);
+                        for (auto& km : kingMoves) {
+                            moves.push_back(current.merge(km));
                         }
-                        Pawn p(isWhite(), newP, board);
-                        p.getCaptureMoves(moves, board, current);
-                    }
-                }
-            }
-            else{
-                if(board.getBoard()[pos.x][pos.y] == WhitePawn || board.getBoard()[pos.x][pos.y] == WhiteKing ){
-                    Position diff = positionDistance(pos);
-                    Position newP = Position(pos.x + diff.x, pos.y + diff.y);
-                    if (board.getBoard()[newP.x][newP.y] == Empty){
-                        countMoves++;
-                        Move tmp = Move(pos_, newP, pos);
-                        current.addStepMove(tmp);
-                        board.makeMove(tmp);
-                        Pawn p(isWhite(), newP, board);
-                        p.getCaptureMoves(moves, board, current);
+                        continue;
                     }
+                    Pawn p(newP, isWhite(), isUser(), board);
+                    p.getCaptureMoves(moves, board, current);
                 }
             }
         }
         
     }
-
-    if(countMoves == 0){
-        moves.push_back(current);
-    }
 }
 
-std::vector<Move> Pawn::getValidMoves(Board &board, bool &mustCapture) {
+std::vector<Move> Pawn::getValidMoves(Board &board, bool mustCapture) {
 
     std::vector<Move> moves;
     Move move;
 
-    mustCapture = true;
     getCaptureMoves(moves, board, move);
 
-    if (moves.size() == 1 && moves[0] == move){
+    if (moves.empty()){
         int d[2] ={1, -1};
-        moves.pop_back();
-        if (this->isWhite() ) {
+        if (isUser()) {
             for(int i = 0 ; i < 2 ; ++i){
-                if(board.getBoard()[pos_.x + 1][pos_.y + d[i]] == Empty){
-                    moves.push_back(Move(pos_,Position(pos_.x + 1, pos_.y +d[i])));
+                Position nP(getPosition().x + d[i], getPosition().y + 1);
+                if(board.getPieceName(nP) == Empty){
+                    moves.push_back(Move(pos_, nP));
                 }
             }
         }
         else {
             for(int i = 0 ; i < 2 ; ++i){
-                if(board.getBoard()[pos_.x - 1][pos_.y + d[i]] == Empty){
-                    moves.push_back(Move(pos_,Position(pos_.x - 1, pos_.y +d[i])));
+                Position nP(getPosition().x + d[i], getPosition().y - 1);
+                if(board.getPieceName(nP) == Empty){
+                    moves.push_back(Move(pos_, nP));
```

Comment:
Jak już się trzyamy getPosition() to chyba powinno być też przy metodzie push_back.